rlinf/runners/online_coding_runner.py

```diff
@@ -0,0 +1,317 @@
+            ncols=620,
+        )
+
+        self.online_router.server_start()
+        self.server_rollout.server_start()
+        self.run_timer.start_time()
+        for _ in range(self.cfg.runner.max_epochs):
+            with self.timer("step"):
+
+                with self.timer("sync_weights"):
+                    self._sync_weights()
+
+                rollout_handle: Handle = self.server_rollout.rollout(
+                    output_channel=self.dataloader_channel,
+                )
+
+                if self.recompute_logprobs:
+                    # Inference prev/ref logprobs
+                    infer_handle: Handle = self.inference.run_inference(
+                        input_channel=self.dataloader_channel,
+                        output_channel=self.inference_channel,
+                        compute_ref_logprobs=self.compute_ref_logprobs,
+                    )
+                    inference_channel = self.inference_channel
+                else:
+                    infer_handle = None
+                    inference_channel = self.dataloader_channel
+
+                # Advantages and returns
+                adv_handle: Handle = self.actor.compute_advantages_and_returns(
+                    input_channel=self.inference_channel,
+                    output_channel=self.actor_channel,
+                )
+
+                # Actor training
+                actor_input_channel = self.actor_channel
+                if self.is_pipeline:
+                    # In pipeline mode, the rollout already contains the advantages and returns
+                    # So the above two steps are in fact no-ops, and we should directly use the inference channel as the input
+                    actor_input_channel = inference_channel
+                actor_handle: Handle = self.actor.run_training(
+                    input_channel=actor_input_channel,
+                )
+
+                metrics = actor_handle.wait()
+                self.global_steps += 1
+
+                run_time_exceeded = self.run_timer.is_finished()
+                _, save_model, is_train_end = check_progress(
+                    self.global_steps,
+                    self.max_steps,
+                    self.cfg.runner.val_check_interval,
+                    self.cfg.runner.save_interval,
+                    1.0,
+                    run_time_exceeded=run_time_exceeded,
+                )
+
+                if save_model:
+                    self._save_checkpoint()
+
+                if is_train_end:
+                    logging.info(
+                        f"Step limit given by max_steps={self.max_steps} reached. Stopping run"
+                    )
+                    return
+
+                if run_time_exceeded:
+                    logging.info(
+                        f"Time limit given by run_timer={self.run_timer} reached. Stopping run"
+                    )
+                    return
+
+                rollout_handle.wait()
```

Comment:
it is a little strange that you call `rollout_handle.wait()` here instead of `actor_handle.wait()`, why?